<span style="color:blue"><font size="6"> **Анализ базы данных сервиса чтения книг по подписке** </span>

**Цель исследования:**
Провести анализ базы данных и чтобы в дальнейшем можно было сформулировать предложение для нового продукта.


**Описание данных:**

- **Таблица 'books'** - содержит данные о книгах
- **Таблица 'authors'** - содержит данные об авторах
- **Таблица 'publishers'** - cодержит данные об издательствах
- **Таблица 'ratings'** - cодержит данные о пользовательских оценках книг
- **Таблица 'reviews'** - cодержит данные о пользовательских обзорах

## Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Загрузка и изучение данных

### Таблица books

In [2]:
books = """ SELECT * FROM books """
con = engine.connect()
books_sql = pd.io.sql.read_sql(sql=text(books), con=con)
display(books_sql.head())
print(' ')
books_sql.info()
print(' ')
print('\033[1m' + 'Пропуски в таблице:' + '\033[0m')
display(books_sql.isna().sum())
print('\033[1m' + 'Количество дубликатов в таблице:' + '\033[0m', books_sql.duplicated().sum())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
 
Пропуски в таблице:


book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

Количество дубликатов в таблице: 0


**В таблице данные о 1000 книгах, пропусков нет, явные дубликаты отсутствуют.**

### Таблица authors

In [3]:
authors = """ SELECT * FROM authors """
con = engine.connect()
authors_sql = pd.io.sql.read_sql(sql=text(authors), con=con)
display(authors_sql.head())
print(' ')
authors_sql.info()
print(' ')
print('\033[1m' + 'Пропуски в таблице:' + '\033[0m')
display(authors_sql.isna().sum())
print('\033[1m' + 'Количество дубликатов в таблице:' + '\033[0m', authors_sql.duplicated().sum())

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
 
Пропуски в таблице:


author_id    0
author       0
dtype: int64

Количество дубликатов в таблице: 0


**В таблице данные о 636 авторов, пропусков нет, явные дубликаты отсутствуют.**

### Таблица publishers

In [4]:
publishers = """ SELECT * FROM publishers """
con = engine.connect()
publishers_sql = pd.io.sql.read_sql(sql=text(publishers), con=con)
display(publishers_sql.head())
print(' ')
publishers_sql.info()
print(' ')
print('\033[1m' + 'Пропуски в таблице:' + '\033[0m')
display(publishers_sql.isna().sum())
print('\033[1m' + 'Количество дубликатов в таблице:' + '\033[0m', publishers_sql.duplicated().sum())

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
 
Пропуски в таблице:


publisher_id    0
publisher       0
dtype: int64

Количество дубликатов в таблице: 0


**В таблице данные о 340 издательствах, пропусков нет, явные дубликаты отсутствуют.**

### Таблица ratings

In [5]:
ratings = """ SELECT * FROM ratings """
con = engine.connect()
ratings_sql = pd.io.sql.read_sql(sql=text(ratings), con=con)
display(ratings_sql.head())
print(' ')
ratings_sql.info()
print(' ')
print('\033[1m' + 'Пропуски в таблице:' + '\033[0m')
display(ratings_sql.isna().sum())
print('\033[1m' + 'Количество дубликатов в таблице:' + '\033[0m', ratings_sql.duplicated().sum())

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
 
Пропуски в таблице:


rating_id    0
book_id      0
username     0
rating       0
dtype: int64

Количество дубликатов в таблице: 0


**В таблице 6456 записи с оценками пользователей, пропусков нет, явные дубликаты отсутствуют.**

### Таблица reviews

In [6]:
reviews = """ SELECT * FROM reviews """
con = engine.connect()
reviews_sql = pd.io.sql.read_sql(sql=text(reviews), con=con)
display(reviews_sql.head())
print(' ')
reviews_sql.info()
print(' ')
print('\033[1m' + 'Пропуски в таблице:' + '\033[0m')
display(reviews_sql.isna().sum())
print('\033[1m' + 'Количество дубликатов в таблице:' + '\033[0m', reviews_sql.duplicated().sum())

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
 
Пропуски в таблице:


review_id    0
book_id      0
username     0
text         0
dtype: int64

Количество дубликатов в таблице: 0


**В таблице 2793 записи с обзорами книг от пользователей, пропусков нет, явные дубликаты отсутствуют.**

## Анализ данных

### Книги, выпущенные после 01.01.2000

In [7]:
query_books = """ 
         SELECT COUNT(book_id)
         FROM books
         WHERE publication_date > '2000-01-01' 
         """
result = pd.io.sql.read_sql(query_books, con=con)
print('После 01.01.2000 выпущено', result.iloc[0,0], 'книг')

После 01.01.2000 выпущено 819 книг


### Количество обзоров и средняя оценка каждой книги

In [8]:
query_reviews = """ 
         SELECT rt.book_id,
                COUNT(review_id) AS cnt_reviews,
                ROUND(AVG(rating),2) AS avg_rating
         FROM ratings rt
         LEFT JOIN reviews r ON r.book_id = rt.book_id AND r.username = rt.username
         GROUP BY rt.book_id
         ORDER BY cnt_reviews DESC 
         """
pd.io.sql.read_sql(query_reviews, con=con)

book_id  cnt_reviews  avg_rating
0        948            7        3.66
1        695            6        4.08
2        299            6        4.29
3        696            6        3.83
4        734            6        4.21
..       ...          ...         ...
995      672            0        5.00
996      808            0        3.00
997      221            0        4.00
998      387            0        4.00
999       83            0        3.67

[1000 rows x 3 columns]

**Максимальное число обзоров по книге - 7, при этом есть и книги, которые не получили ни одного обзора, однако, оценку пользователи все же оставили.**

### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
query_publisher = """ 
         SELECT p.publisher,
                COUNT(b.book_id) AS cnt_books
         FROM books b
         JOIN publishers p ON b.publisher_id=p.publisher_id  
         WHERE b.num_pages > 50
         GROUP BY p.publisher
         ORDER BY cnt_books DESC
         LIMIT 1 
         """
pd.io.sql.read_sql(query_publisher, con=con)

publisher  cnt_books
0  Penguin Books         42

**Penguin Books является издательством, которое выпустило максимальное кол-во книг толще 50 страниц.**

### Автор с самой высокой средней оценкой книг (учитываются только книги с 50 и более оценками)

In [22]:
query_author = """ 
         SELECT a.author,
                ROUND(AVG(r.avg_rating),2)
         FROM (SELECT book_id,
                      COUNT(rating) AS cnt_rating,
                      ROUND(AVG(rating),2) AS avg_rating
               FROM ratings
               GROUP BY book_id
               HAVING COUNT(rating) >= 50) r
         JOIN books b ON r.book_id=b.book_id
         JOIN authors a ON b.author_id=a.author_id
         GROUP BY a.author
         ORDER BY AVG(r.avg_rating) DESC
         LIMIT 1
         """
pd.io.sql.read_sql(query_author, con=con)

author  round
0  J.K. Rowling/Mary GrandPré   4.29

**Джоан Кэтлин Роулинг получила самую высокую оценку от пользователей - 4,29.**

### Среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [11]:
query_avg_reviews = """ 
         SELECT AVG(cnt_reviews) 
         FROM (SELECT r.username,
                      COUNT(review_id) AS cnt_reviews
               FROM reviews r
               WHERE username IN (SELECT username
                                  FROM ratings
                                  GROUP BY username
                                  HAVING COUNT(rating_id) > 48) 
               GROUP BY r.username) t
         """
pd.io.sql.read_sql(query_avg_reviews, con=con)

avg
0  24.0

**В среднем пользователи, которые ставят более 48 оценок, оставляют 24 обзора на книги.**

## Вывод

В базе данных сервиса содержится информация о 1000 книгах, данные без пропусков и явных дубликатов.

- Было выпущено 819 книг после 01.01.2000 - 82% от выборки
- Замечено 6 книг, которые не получили обзоры от пользователей, но получили оценки
- Максимальное количество обзоров на книгу составило 7
- В среднем пользователи, которые ставят более 48 оценок, оставляют 24 обзора на книги
- Издательство, которое выпустило максимальное число книг толще 50 страниц - Penguin Books (42 книги)
- Джоан Кэтлин Роулинг, как автор, получила самую высокую оценку от пользователей - 4,29